### PR TITLE
fix: DOT round-trip for block-form parallel branches (#76)

### LINKED ISSUE #76: bug: DOT export silently drops all block-form parallel branches (per-branch model/provider/fidelity + targets)
## Summary

`export.ExportDOT` does not serialize block-form parallel branches at all. `applyParallelAttrs` (export/dot.go:377) writes only `cfg.Targets` (the inline fan-out form) and ignores `cfg.Branches` entirely. Any `ParallelConfig` authored in **block form** loses **all** per-branch configuration when exported to DOT — not just `model`/`provider`/`fidelity`, but each branch's `target` too.

This is the standalone data-loss bug split out from the discussion in #58 (which is scoped to adding `BranchConfig.ToolAccess`). Verified against `main` at v0.34.0.

## Impact / affected commands

`ExportDOT` is reached by:
- `dippin export` (`cmd/dippin/cmd_export.go:47`)
- `dippin simulate` DOT visualization (`cmd/dippin/cmd_simulate.go:142`)

Severity depends on form:
- **Block form with inline targets present** → per-branch `model`/`provider`/`fidelity` silently dropped (attribute loss).
- **Block-form-only node (no inline `Targets`)** → exported DOT parallel node has *no fan-out targets at all* — the fan-out is structurally broken, not merely lossy.

The loss is silent: no warning, no error.

## Why the `.dip`→`.dip` path is fine

The native formatter (`formatter/format.go:241,258`) and simulation (`simulate/parallel.go:39`) both read `cfg.Branches` correctly. The gap is specific to the **DOT export path** — only `export/dot.go` is blind to branches.

## Repro

1. Author a parallel node in block form with per-branch model/provider/fidelity (e.g. the pattern in `examples/` that uses block-form branches).
2. `dippin export <file>.dip` to DOT.
3. Observe the emitted `parallel` node has no `targets=` reflecting the branches (and no per-branch attrs).

## Fix sketch

`applyParallelAttrs` needs to emit `cfg.Branches`. Options:
- Encode per-branch attrs into DOT (needs a DOT-representable encoding — branches carry target+model+provider+fidelity, so a flat `targets=` is insufficient; likely per-branch sub-attributes or a structured encoding the migrate-side parser can read back).
- At minimum, flatten branch `Target`s into `targets=` so the fan-out structure survives, and `log`/warn that per-branch attrs are dropped (parallels the documented DOT-lossiness pattern for `command_file:` in #69).

Whichever encoding is chosen must round-trip through the migrate-side DOT parser (`migrate/migrate.go`), which currently has no branch-reading logic at all.

## Acceptance

- Block-form parallel node survives `dippin export` → DOT → (migrate) → `.dip` without losing targets.
- Either per-branch model/provider/fidelity round-trip, or the loss is explicitly warned (not silent).
- Test added covering a block-form-only parallel node through the DOT round-trip.

### PR DESCRIPTION
**Pull request by clintecker**
## Summary
Block-form `parallel` nodes now round-trip losslessly through `dippin export` → DOT → `migrate`. Previously the DOT path wrote only `targets=` and dropped per-branch `model`/`provider`/`fidelity` entirely (and could lose fan-out for block-form-only IR). Fixes #76.

- **Export** emits a new `branches=` node attribute — each branch as `;`-joined `key=value` tokens (`target` plus any of `model`/`provider`/`fidelity`), branches joined by `,`, reserved chars `% , ; =` percent-encoded — mirroring the existing `steer_context` encoder. `targets=` is always co-emitted for backward compatibility.
- **Migrate** decodes `branches=` back into `[]ir.BranchConfig`; branch order is the source of truth for `Targets` (reproducing the parser invariant `Targets[i] == Branches[i].Target`). Target-less/malformed tokens are dropped, not silently corrupted.
- Three correctness fixes the per-task review surfaced:
  - `inferParallelTargets` no longer discards `Branches` when backfilling targets from edges.
  - The `--check` parity comparison (`compareParallelConfigs`) now compares `Branches` field-by-field — previously it compared only `Targets`, which is exactly why the loss was silent.
  - `resolveStartExitKind` now recovers `NodeParallel` for a parallel node used as the workflow start/exit (via `hasParallelAttrs`), so the canonical `start: split` pattern round-trips.

## Unblocks
Unblocks #58 (per-branch `tool_access` override): per-branch attributes can now survive DOT, and the `k=v` encoding makes adding a `tool_access` key a one-line change.

## Design / review
Brainstormed and spec'd before implementation, then the design was audited by a 4-reviewer expert panel (round-trip correctness, codebase consistency, complexity, backward-compat); their findings (the parity blind spot and the inference clobber especially) are folded into the spec and the code. Spec: `docs/superpowers/specs/2026-05-28-issue-76-dot-parallel-branches-design.md`.

## Test Plan
- [x] `just check` green (build, vet, fmt, test-race, complexity ≤5/≤7, validate-examples, tree-sitter)
- [x] Export/migrate unit tests for `branches=` encode/decode
- [x] Reserved-char (`% , ; =`) round-trip test
- [x] Malformed-input tolerance test
- [x] `inferParallelTargets` Branches-preservation test
- [x] Parallel-as-start-node recovery test
- [x] Full `.dip` → export → migrate acceptance round-trip (block-form-only)

Note: CHANGELOG intentionally untouched — it's updated at tag time per CLAUDE.md.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/2389-research/codesmith/dippin-lang/pr/80"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1782663011&installation_id=131176874&pr_number=80&repository=2389-research%2Fdippin-lang&return_to=https%3A%2F%2Fgithub.com%2F2389-research%2Fdippin-lang%2Fpull%2F80&signature=8191b53512e21b4e40e1a2c95534800c7519ffbe11a8b642ee403b0ff3d20e40"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **New Features**
  * Block-form parallel nodes now export a branches= DOT attribute so per-branch overrides (target/model/provider/fidelity) round-trip losslessly while still emitting targets= for compatibility.

* **Bug Fixes**
  * Parallel node kind, branch ordering and branch metadata are preserved during DOT export/migrate; malformed or target-less branches are dropped.

* **Documentation**
  * Added design, plan, and node docs describing block-form branch encoding, ordering, and percent-encoding of reserved characters.

* **Tests**
  * Added unit and acceptance tests for encoding/decoding, round-trips, malformed branches, parity checks, and reserved-character edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/2389-research/dippin-lang/pull/80?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: export/dot.go [hunk 1]
```diff
@@ -373,11 +373,93 @@ func applySubgraphAttrs(attrs map[string]string, cfg ir.SubgraphConfig) {
...
+// (the escape char itself, replaced first to avoid double-encoding). This is a
+// superset of steerContextEncoder — branches add ';'. The outer DOT-quote layer
+// (dotQuote) handles '"', '\', and newlines, so those are intentionally not
+// percent-encoded here.
+var branchEncoder = strings.NewReplacer(
+	"%", "%25",
+	",", "%2C",
+	";", "%3B",
+	"=", "%3D",
+)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply export/dot.go:461]: Fixed in f94c11b. The branch encoder now percent-encodes `\` (→`%5C`) in addition to `% , ; =`, so a literal backslash followed by n/l/r can no longer survive the DOT-quote layer as a `\n`/`\l`/`\r` escape and be decoded to a newline by the migrate lexer. `TestExportDOT_Parallel_BranchesReservedCharsRoundTrip` now runs through full `Migrate()` (not just `parseBranches`) and includes a `back\nslash` value to cover exactly this path.

File: docs/nodes.md [hunk 2]
```diff
@@ -424,6 +424,34 @@ You need edges connecting the parallel node to its targets and the targets to th
...
+
+Each `branch:` entry declares a fan-out target (equivalent to an inline `-> fast, accurate`) and attaches per-branch overrides for `model`, `provider`, and `fidelity`. The fan-in node must still list the same target IDs.
+
+#### DOT mapping
+
+Block-form parallels export a `branches=` node attribute alongside the standard `targets=` attribute. Each branch is serialised as `;`-joined `key=value` tokens (`target` plus any of `model`/`provider`/`fidelity`), with branches joined by `,`:
+
+```
+branches="target=fast;model=claude-haiku-4-5;provider=anthropic;fidelity=summary,target=accurate;model=claude-opus-4-7;provider=anthropic;fidelity=full"
+```
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/nodes.md]: Fixed in f94c11b — the DOT example fence is now ```text (and the section was also moved under Parallel Nodes).

File: docs/superpowers/specs/2026-05-28-issue-76-dot-parallel-branches-design.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+Each branch is encoded as uniform `k=v` tokens joined by `;`:
+
+- `target` is always present (first).
+- `model` / `provider` / `fidelity` only when non-empty.
+
+Branches are joined by `,`, in slice order (order is semantically meaningful —
+it maps positionally to targets). Example node attribute:
+
+```
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/superpowers/specs/2026-05-28-issue-76-dot-parallel-branches-design.md]: Fixed in f94c11b — tagged the fence ```dot.

File: migrate/migrate.go [hunk 4]
```diff
@@ -582,10 +599,101 @@ func applyToolOutputLimitAttr(cfg *ir.ToolConfig, attrs map[string]string) error
...
+	}
+}
+
+// branchDecoder reverses branchEncoder from export/dot.go.
+var branchDecoder = strings.NewReplacer(
+	"%25", "%",
+	"%2C", ",",
+	"%3B", ";",
+	"%3D", "=",
+)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply migrate/migrate.go:693]: Not a bug — declining the reorder. Go's `strings.NewReplacer` is single-pass and never re-scans inserted output, and the `%25/%2C/%3B/%3D/%5C` patterns are mutually exclusive (none is a prefix of another), so replacement order is irrelevant: a literal `%2C` encodes to `%252C` and decodes back to `%2C`, not `,`. f94c11b adds a regression test (`Provider: "%2C"` in the reserved-chars round-trip through full `Migrate()`) and a code comment documenting this. Reordering would falsely imply a fix was needed.

File: migrate/parity.go [hunk 5]
```diff
@@ -349,12 +349,36 @@ func compareParallelConfigs(id, path string, ac ir.ParallelConfig, bCfg interfac
...
+
+// branchesKey produces a stable comparison key for a branch slice (order
+// matters — it maps positionally to targets).
+func branchesKey(branches []ir.BranchConfig) string {
+	parts := make([]string, len(branches))
+	for i, b := range branches {
+		parts[i] = fmt.Sprintf("%s|%s|%s|%s", b.Target, b.Model, b.Provider, b.Fidelity)
+	}
+	return strings.Join(parts, ",")
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply migrate/parity.go:379]: Fixed in f94c11b. `compareParallelBranches` now compares element-wise (`ac.Branches[i] != bc.Branches[i]` — `BranchConfig` is an all-string comparable struct) and `branchesKey` is removed, so values containing `|`/`,` no longer collide. Used `!=` rather than the suggested 4-way `||` loop because the latter would push the function to cyclo 7, over the repo's ≤5 cap.

File: migrate/roundtrip_test.go [hunk 6]
```diff
@@ -337,3 +337,104 @@ func TestRoundtripPreservesToolAccess(t *testing.T) {
...
+	got := parseBranches(attrVal)
+	if len(got) != len(original) {
+		t.Fatalf("branches = %d, want %d", len(got), len(original))
+	}
+	for i, want := range original {
+		if got[i] != want {
+			t.Errorf("branch[%d] round-trip: got %+v want %+v", i, got[i], want)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply migrate/roundtrip_test.go:405]: Declining. These tests verify the DOT export→migrate round-trip (Targets/Branches preservation); the parse→lint integration contract is `TestLintExamples`'s job over the `examples/*.dip` corpus. Running lint inside a DOT round-trip test conflates concerns and would not exercise the round-trip behavior under test.

File: migrate/roundtrip_test.go [hunk 7]
```diff
@@ -337,3 +337,104 @@ func TestRoundtripPreservesToolAccess(t *testing.T) {
...
+	want := []ir.BranchConfig{
+		{Target: "fast", Model: "claude-haiku-4-5", Provider: "anthropic", Fidelity: "summary"},
+		{Target: "accurate", Model: "claude-opus-4-7", Provider: "anthropic", Fidelity: "full"},
+	}
+	for i := range want {
+		if cfg.Branches[i] != want[i] {
+			t.Errorf("branch[%d]: got %+v want %+v", i, cfg.Branches[i], want[i])
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply migrate/roundtrip_test.go:456]: The complexity caps are enforced by `just complexity`, which greps out `_test.go` — they do not apply to test functions. That said, f94c11b extracts `assertNodeParallelConfig` and `assertBranchesEqual` helpers (shared by both round-trip tests), which reduces the inline decision points anyway.

File: migrate/roundtrip_test.go [hunk 8]
```diff
@@ -337,3 +337,104 @@ func TestRoundtripPreservesToolAccess(t *testing.T) {
...
+	}
+	dot := export.ExportDOT(w, export.ExportOptions{})
+	got, err := Migrate(dot)
+	if err != nil {
+		t.Fatalf("migrate: %v", err)
+	}
+	cfg, ok := got.Node("split").Config.(ir.ParallelConfig)
+	if !ok {
+		t.Fatalf("config type = %T, want ParallelConfig", got.Node("split").Config)
+	}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply migrate/roundtrip_test.go]: Fixed in f94c11b via the `assertNodeParallelConfig` helper, which checks `w.Node(id) == nil` and `t.Fatalf`s with a clear message before the type assertion.

File: migrate/parity.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (clintecker) [reply migrate/parity.go:379]: Fixed in f94c11b — `compareParallelBranches` now compares branches element-wise via `!=` on the comparable `BranchConfig` struct; `branchesKey` removed. Nil-vs-empty is covered by the leading length check.

File: docs/nodes.md [hunk 9]
```diff
@@ -424,6 +424,34 @@ You need edges connecting the parallel node to its targets and the targets to th
 
 Some of these edges may be auto-generated by the parser if omitted.
 
+### Block Form
+
+Use block form when branches need different models, providers, or fidelity levels:
+
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/nodes.md]: Fixed in f94c11b — the Block Form section is moved under `## Parallel Nodes` (after How It Works), so it's no longer under Fan-In Nodes.

File: docs/nodes.md [hunk 10]
```diff
@@ -424,6 +424,34 @@ You need edges connecting the parallel node to its targets and the targets to th
...
+      model: claude-opus-4-7
+      provider: anthropic
+      fidelity: full
+```
+
+Each `branch:` entry declares a fan-out target (equivalent to an inline `-> fast, accurate`) and attaches per-branch overrides for `model`, `provider`, and `fidelity`. The fan-in node must still list the same target IDs.
+
+#### DOT mapping
+
+Block-form parallels export a `branches=` node attribute alongside the standard `targets=` attribute. Each branch is serialised as `;`-joined `key=value` tokens (`target` plus any of `model`/`provider`/`fidelity`), with branches joined by `,`:
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/nodes.md]: Fixed in f94c11b — changed to "serialized".

File: docs/nodes.md [hunk 11]
```diff
@@ -361,6 +361,34 @@ parallel <ID> -> <target1>, <target2>[, <target3>, ...]
...
+
+#### DOT mapping
+
+Block-form parallels export a `branches=` node attribute alongside the standard `targets=` attribute. Each branch is serialized as `;`-joined `key=value` tokens (`target` plus any of `model`/`provider`/`fidelity`), with branches joined by `,`:
+
+```text
+branches="target=fast;model=claude-haiku-4-5;provider=anthropic;fidelity=summary,target=accurate;model=claude-opus-4-7;provider=anthropic;fidelity=full"
+```
+
+The reserved characters `%`, `,`, `;`, `=`, and `\` are percent-encoded in keys and values (the same convention used by `steer_context`), so branch metadata round-trips losslessly through `migrate`.
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/nodes.md]: Fixed in ed1c2bd — reworded to "the same percent-encoding approach as `steer_context`, though branches reserve a few more characters" so it no longer implies an identical reserved set.

File: docs/superpowers/specs/2026-05-28-issue-76-dot-parallel-branches-design.md [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+  targets="fast,accurate",
+  branches="target=fast;model=claude-haiku-4-5;provider=anthropic;fidelity=summary,target=accurate;model=claude-opus-4-7;provider=anthropic;fidelity=full"];
+```
+
+### Reserved-character safety
+
+The four structural characters `%`, `,`, `;`, `=` are percent-encoded inside keys
+and values (`%`→`%25`, `,`→`%2C`, `;`→`%3B`, `=`→`%3D`), with `%` replaced first
+on encode. This is a **superset** of the `steer_context` encoder (which reserves
+only `% , =`) because `;` is the new within-branch field separator.
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/superpowers/specs/2026-05-28-issue-76-dot-parallel-branches-design.md]: Fixed in ed1c2bd — the spec's reserved-character section now lists `\`→`%5C` and explains it prevents a literal backslash + n/l/r from being interpreted as a DOT escape by the migrate lexer.

File: docs/superpowers/specs/2026-05-28-issue-76-dot-parallel-branches-design.md [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+We do **not** reuse `steerContextEncoder` — adding `;` to it would change the
+`steer_context` wire format and break its round-trip tests. Instead each package
+gets a dedicated `branchEncoder`/`branchDecoder` `strings.Replacer` pair, mirroring
+the steer pattern (encoder in `export`, decoder in `migrate`; the two packages
+cannot share private helpers and already duplicate the steer encoder/decoder
+across the boundary — this is the established, accepted norm).
+
+The outer DOT-quote layer (`dotQuote` / lexer `readString`) handles `"`, `\`, and
+newlines, so those are intentionally **not** in the percent set. A code comment
+will note this so a future editor doesn't assume the percent set is self-sufficient.
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/superpowers/specs/2026-05-28-issue-76-dot-parallel-branches-design.md]: Fixed in ed1c2bd — the paragraph now states that only `"` and newlines rely on the DOT-quote layer, and that `\` is percent-encoded specifically to defeat DOT escape-sequence interpretation.

File: docs/superpowers/plans/2026-05-28-issue-76-dot-parallel-branches.md [hunk 14]
```diff
@@ -0,0 +1,811 @@
...
+// (the escape char itself, replaced first to avoid double-encoding). This is a
+// superset of steerContextEncoder — branches add ';'. The outer DOT-quote layer
+// (dotQuote) handles '"', '\', and newlines, so those are intentionally not
+// percent-encoded here.
+var branchEncoder = strings.NewReplacer(
+	"%", "%25",
+	",", "%2C",
+	";", "%3B",
+	"=", "%3D",
+)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/superpowers/plans/2026-05-28-issue-76-dot-parallel-branches.md]: Fixed in ed1c2bd — the plan's `branchEncoder` snippet now includes `"\\", "%5C"` with an updated comment explaining why.

File: docs/superpowers/plans/2026-05-28-issue-76-dot-parallel-branches.md [hunk 15]
```diff
@@ -0,0 +1,811 @@
...
+
+// encodeBranchToken percent-encodes the reserved delimiter characters in a key
+// or value so the round-trip through DOT → migrate stays lossless.
+func encodeBranchToken(s string) string {
+	if !strings.ContainsAny(s, ",;=%") {
+		return s
+	}
+	return branchEncoder.Replace(s)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/superpowers/plans/2026-05-28-issue-76-dot-parallel-branches.md:184]: Fixed in ed1c2bd — the plan's `encodeBranchToken` snippet now uses `strings.ContainsAny(s, ",;=%\\")`, matching the implementation. (Also updated the plan's `branchDecoder` snippet to include `"%5C", "\\"`.)
